File: notes/midnight-city/decide.py
#!/usr/bin/env python3
# Decide one agent's next action from inventory/progression/needs JSON.
# Priority: 1) stay fed (sell goods to afford food if broke)  2) shed load when
# overburdened  3) complete a ready contract  4) craft once unlocked  5) sell surplus
# 6) workers funnel crystal to the treasury, keeping a food buffer  7) grind XP.
import sys, json, os

inv_f, prog_f, needs_f, mainskill, good, batch, sellat, isfloyd = sys.argv[1:9]
batch, sellat, isfloyd = int(batch), int(sellat), isfloyd == "1"

MEAL_COST = 23        # Central Smoothies Matcha Outlet enforces multiples of 23 (listing says 20);
                      # one smoothie restored 20 hunger in play (content says 46). Cheapest per point.
FOOD_BUFFER = 70      # workers keep enough for three smoothies before funnelling to Floyd

def load(f):
    try:
        return json.load(open(f))
    except Exception:
        return {}

# Food = any item with hungerRestore > 0 in the game content dump (100+ items).
CONTENT = os.path.expanduser("~/.midnight-city/notes/game-content.json")
FOOD_IDS = {i["id"] for i in load(CONTENT).get("items", [])
            if ((i.get("consumable") or {}).get("hungerRestore") or 0) > 0}
FOOD_HINTS = ("food", "fish", "meat", "ration", "brew", "stew", "blend", "bun", "soup")

inv_doc = load(inv_f)
inv = inv_doc.get("inventory", {}) or {}
load_state = (inv_doc.get("load", {}) or {}).get("state", "")
prog = load(prog_f)
needs = load(needs_f)

crystal = int(inv.get("crystal", 0) or 0)
goods = int(inv.get(good, 0) or 0)
sellable = (goods // batch) * batch
skills = prog.get("skills", {}) or {}
lvl = skills.get(mainskill, {}).get("level", "?")
xp = skills.get(mainskill, {}).get("xp", 0)
caps = prog.get("capabilities", {}) or {}
recipes = caps.get("recipes", []) or []
contracts = caps.get("contracts", []) or []
hunger = (needs.get("hunger", {}) or {}).get("state", "normal")

def is_food(k):
    return k in FOOD_IDS or (not FOOD_IDS and any(h in k for h in FOOD_HINTS))
foods = [k for k in inv if is_food(k) and int(inv.get(k, 0) or 0) > 0]

def has_all(reqs):
    return all(int(inv.get(r.get("itemId"), 0)) >= int(r.get("quantity", 1)) for r in (reqs or []))

action, detail = "WORK", ""

# crystal transfers have a weekly allowance; run-crew.sh drops a flag file when it runs out
import time
flag = os.path.expanduser(f"~/.midnight-city/tmp/no-send-{os.environ.get('NM', '')}")
send_blocked = os.path.exists(flag) and time.time() - os.path.getmtime(flag) < 7 * 86400

# 0) a trade walks the agent to the merchant first; a new action would cut it short
active = (inv_doc.get("agent", {}) or {}).get("activeAction") or {}
busy_elsewhere = active and active.get("kind") not in ("engage", "perform_job", None)

# 1) self-sustain: eat, else buy (two meals if affordable), else sell goods to afford it
if busy_elsewhere:
    action, detail = "WAIT", active.get("kind", "")
elif hunger not in ("normal", "full", "", None):
    if foods:
        action = "EAT"
    elif crystal >= MEAL_COST:
        action, detail = "BUYFOOD", str(min(crystal // MEAL_COST, 2) * MEAL_COST)
    elif sellable > 0:
        action, detail = "SELL", str(sellable)
    else:
        action = "WORK"
# 2) overburdened agents work at a fraction of speed: sell the trade good first
elif load_state == "overburdened" and sellable > 0:
    action, detail = "SELL", str(sellable)
else:
    # 3) deliver a ready, uncompleted contract
    ready_contract = None
    for c in contracts:
        if c.get("completed"):
            continue
        reqs = c.get("requirements")
        if reqs is not None:
            if has_all(reqs):
                ready_contract = c["contractId"]; break
        else:
            # requirements not inlined: assume the trade good satisfies a same-skill contract
            if c.get("skill") == mainskill and goods >= 1:
                ready_contract = c["contractId"]; break
    if ready_contract:
        action, detail = "DELIVER", ready_contract
    else:
        # 4) craft an unlocked recipe whose inputs we hold
        craftable = None
        for r in recipes:
            if r.get("inputs") and has_all(r["inputs"]):
                craftable = r["id"]; break
            # some recipe entries expose craftableBatches>0 as "ready now"
            if not r.get("inputs") and int(r.get("craftableBatches", 0)) > 0:
                craftable = r.get("recipeId") or r.get("id"); break
        # crafting must not starve selling: sell first once the pile passes 5x the threshold
        if goods >= sellat * 5 and sellable > 0:
            action, detail = "SELL", str(sellable)
        elif craftable:
            action, detail = "CRAFT", craftable
        # 5) sell surplus
        elif goods >= sellat and sellable > 0:
            action, detail = "SELL", str(sellable)
        # 6) workers funnel surplus to treasury, keeping a food buffer so they never starve
        elif (not isfloyd) and crystal > FOOD_BUFFER and not send_blocked:
            action, detail = "SEND", str(crystal - FOOD_BUFFER)
        else:
            action = "WORK"

n_open = sum(1 for c in contracts if not c.get("completed"))
status = f"L{lvl} {mainskill} xp{xp} | contracts_open={n_open} recipes={len(recipes)} | hunger={hunger} | crystal={crystal} | {good}={goods} foods={len(foods)} load={load_state or '-'}"
print(f"{action}|{detail}|{status}")

File: worker/src/decide.js
// One agent's next action, ported from notes/midnight-city/decide.py.
// Priority: 0) do not interrupt a trade walk  1) stay fed: eat early, fish when out of food,
// buy food only as a fallback  1a) keep a small fish stock  1b) buy the best tool on sale
// 1c) Floyd funds a worker's tool  2) shed load when overburdened  3) deliver a ready contract
// 4) sell a big pile  5) work a contract, and self-supply a tool nobody sells (gather, craft,
// train); the two alternate by round so neither starves the other
// 6) sell surplus  7) workers fund Floyd, keeping a food buffer  8) work.
// Nothing is crafted except for a tool plan, a contract or food, so no surplus builds up.
import { FOOD_IDS } from "./food-ids.js";
import { TOOLS } from "./tools.js";
import { toolToCraft, nextStep, trainStep, toolSkills, heldBonus } from "./planner.js";
import { deliverable, pursue } from "./contract-plan.js";
import { CONTRACTS } from "./contracts.js";

// Cheapest food per hunger point. The server enforced 23 crystal per smoothie while
// `merchants` listed 20 (FINDINGS 15); the tick updates this from failure events.
export const FOOD_MERCHANT = "Central Smoothies Matcha Outlet";
export const DEFAULT_MEAL_COST = 23;
const FOOD_BUFFER_MEALS = 3;

// Fishing is free: Canal Eddy needs no rod at fishing level 1, and each gather yields one
// fish, one river eel and one canal carp. One fish restored about 50 hunger in play.
// Never cook: cooked_fish cannot be eaten even though the game lists it as consumable
// (FINDINGS 22), so cooking destroys the crew's food.
const FISHING_SOURCE = "canal_eddy";
const EAT_AT = 45;      // eat before "hungry"; a fish restored about 50 in play
const FOOD_STOCK = 3;   // about a day of fish at ~100 hunger a day

const count = (inv, id) => Number(inv?.[id] ?? 0) || 0;
const hasAll = (inv, reqs) => (reqs ?? []).every((r) => count(inv, r.itemId) >= Number(r.quantity ?? 1));

// agent: roster entry. docs: { inventory, progression, needs } responses.
// opts: { mealCost, sendBlocked, treasuryId, round, toolOffers: [{ itemId, merchantName, price }],
//         fundRequests: { workerId: crystal } (treasury only), inedible: [itemId],
//         movedFor: contractId we have already travelled to, delivered: [contractId],
//         skipContracts: { contractId: until } goals that stopped making progress,
//         buyers: [{ itemId, merchantName, batch }] merchants that buy items for crystal }
export function decide(agent, docs, opts) {
  const invDoc = docs.inventory ?? {};
  const inv = invDoc.inventory ?? {};
  const loadState = invDoc.load?.state ?? "";
  const caps = docs.progression?.capabilities ?? {};
  const recipes = caps.recipes ?? [];
  const contracts = caps.contracts ?? [];
  const skill = docs.progression?.skills?.[agent.skill] ?? {};
  const hunger = docs.needs?.hunger?.state ?? "normal";
  const hungerValue = Number(docs.needs?.hunger?.value ?? 0);
  const fishingNodes = (caps.sources ?? [])
    .find((src) => src.sourceId === FISHING_SOURCE && !src.failureReason)?.availableNodeIds ?? [];

  const mealCost = opts.mealCost || DEFAULT_MEAL_COST;
  const crystal = count(inv, "crystal");
  const goods = count(inv, agent.good);
  const sellable = Math.floor(goods / agent.batch) * agent.batch;
  const inedible = new Set(opts.inedible ?? []);
  const foods = Object.keys(inv).filter((id) => FOOD_IDS.has(id) && !inedible.has(id) && count(inv, id) > 0);
  const foodStock = foods.reduce((sum, id) => sum + count(inv, id), 0);

  const status = `L${skill.level ?? "?"} ${agent.skill} xp${skill.xp ?? 0} | hunger=${hunger}(${hungerValue}) | ` +
    `crystal=${crystal} | ${agent.good}=${goods} food=${foodStock} load=${loadState || "-"}`;
  const result = (label, action) => ({ label, action, status });

  const sell = () => result(`SELL ${sellable}`, {
    kind: "trade", merchantName: agent.merchant, itemId: agent.good, quantity: sellable,
  });
  const work = () => result("WORK", { kind: "perform_job" });
  const fish = () => {
    const nodeId = fishingNodes[(opts.round ?? 0) % fishingNodes.length];
    return result(`FISH ${nodeId}`, { kind: "gather", nodeId });
  };

  // 0) a trade walks the agent to the merchant first; a new action would cut it short
  const active = invDoc.agent?.activeAction;
  if (active && !["engage", "perform_job"].includes(active.kind)) {
    return result(`WAIT ${active.kind}`, null);
  }

  // 1) eat early; when out of food, fish; buy (or sell to afford) only if fishing is unavailable
  const hungry = !["normal", "full", ""].includes(hunger);
  if (foods.length && (hungry || hungerValue >= EAT_AT)) return result("EAT", { kind: "eat" });
  if (hungry) {
    if (fishingNodes.length) return fish();
    if (crystal >= mealCost) {
      const quantity = Math.min(Math.floor(crystal / mealCost), 2) * mealCost;
      return result(`BUYFOOD ${quantity}`, {
        kind: "trade", merchantName: FOOD_MERCHANT, itemId: "crystal", quantity,
      });
    }
    return sellable > 0 ? sell() : work();
  }

  // 1a) keep a small fish stock so nobody ever needs to buy food
  if (foodStock < FOOD_STOCK && fishingNodes.length) return fish();

  // 1b) buy the best tool on sale for a useful skill; if we cannot afford it, ask the treasury
  const skills = docs.progression?.skills ?? {};
  const wanted = bestToolOnSale(agent, inv, skills, opts.toolOffers ?? []);
  const spendable = crystal - mealCost;
  const fundRequest = wanted && wanted.price > spendable ? wanted.price - spendable : 0;
  const withFund = (r) => ({ ...r, fundRequest });
  if (wanted && !fundRequest) {
    return result(`BUYTOOL ${wanted.itemId} ${wanted.price}`, {
      kind: "trade", merchantName: wanted.merchantName, itemId: "crystal", quantity: wanted.price,
    });
  }

  // 1c) the treasury covers a worker's shortfall for a tool on sale, keeping its own buffer
  if (agent.isTreasury && !opts.sendBlocked) {
    for (const [workerId, amount] of Object.entries(opts.fundRequests ?? {})) {
      if (crystal - mealCost * FOOD_BUFFER_MEALS >= amount) {
        return result(`FUND ${workerId.slice(-6)} ${amount}`, { kind: "crystal_transfer", recipientAgentId: workerId, quantity: amount });
      }
    }
  }

  // 2) overburdened agents work at a fraction of speed
  if (loadState === "overburdened") {
    if (sellable > 0) return withFund(sell());
    // Nothing of our own trade good: shift whatever else a merchant will take.
    const spare = (opts.buyers ?? [])
      .map((b) => ({ ...b, qty: Math.floor(count(inv, b.itemId) / b.batch) * b.batch }))
      .filter((b) => b.qty > 0)
      .sort((a, b) => b.qty - a.qty)[0];
    if (spare) {
      return withFund(result(`DUMP ${spare.itemId} ${spare.qty}`, {
        kind: "trade", merchantName: spare.merchantName, itemId: spare.itemId, quantity: spare.qty,
      }));
    }
  }

  // 3) deliver a contract whose goods we hold (the API only lists it once we do)
  // Progression is read once per tick, so a contract delivered this tick would still look
  // outstanding to the later rounds; our own record of deliveries covers the gap.
  const completed = [...(docs.progression?.completedContractIds ?? []), ...(opts.delivered ?? [])];
  // deliver_contract does not walk the agent to the contract area (FINDINGS 23), so move
  // first, then deliver on a later round once we are standing there.
  const ready = deliverable(inv, skills, completed, FOOD_STOCK);
  if (ready) {
    if (opts.movedFor === ready) {
      return withFund(result(`DELIVER ${ready}`, { kind: "deliver_contract", contractId: ready }));
    }
    const area = CONTRACTS[ready].area;
    return withFund({
      ...result(`GOTO ${area} for ${ready.slice(0, 30)}`, { kind: "move_to", destination: { areaId: area } }),
      movingFor: ready,
    });
  }

  // 4) crafting must not starve selling: sell first once the pile passes 5x the threshold
  if (goods >= agent.sellAt * 5 && sellable > 0) return withFund(sell());

  // 5) a contract step and a tool step, alternating by round so neither starves the other
  const ctx = { inv, skills, nodes: nodesBySource(caps), round: opts.round, foodFloor: FOOD_STOCK, skip: opts.skipContracts ?? {} };
  const toolPlan = () => {
    const soldIds = new Set((opts.toolOffers ?? []).map((o) => o.itemId));
    const toolId = toolToCraft(agent, inv, skills, soldIds);
    if (!toolId) return null;
    const step = nextStep(toolId, 1, ctx, 0, [toolId]);
    return step.act ?? (step.blocked ? trainStep(toolId, step.blocked, ctx) : null);
  };
  const contractPlan = () => pursue(ctx, completed);
  const order = (opts.round ?? 1) % 3 === 0 ? [toolPlan, contractPlan] : [contractPlan, toolPlan];
  for (const plan of order) {
    const chosen = plan();
    if (chosen) return withFund({ ...result(chosen.label, chosen.action), goal: chosen.goal });
  }

  // 6) sell surplus
  if (goods >= agent.sellAt && sellable > 0) return withFund(sell());

  // 7) workers fund the treasury, keeping a food buffer (sends have a weekly allowance)
  const buffer = mealCost * FOOD_BUFFER_MEALS;
  if (!agent.isTreasury && crystal > buffer && !opts.sendBlocked && !fundRequest) {
    const quantity = crystal - buffer;
    return result(`SEND ${quantity}`, { kind: "crystal_transfer", recipientAgentId: opts.treasuryId, quantity });
  }

  return withFund(work());
}

const nodesBySource = (caps) => Object.fromEntries((caps.sources ?? [])
  .filter((src) => !src.failureReason)
  .map((src) => [src.sourceId, src.availableNodeIds ?? []]));

// The best tool on sale in a useful skill that the agent can use now and does not already
// beat, regardless of price (the caller decides between buying and asking for funds).
function bestToolOnSale(agent, inv, skills, offers) {
  const level = (skill) => Number(skills?.[skill]?.level ?? 0);
  return offers
    .filter((o) => {
      const t = TOOLS[o.itemId];
      return t && toolSkills(agent, skills).includes(t.skill) && t.level <= level(t.skill) && t.bonus > heldBonus(inv, t.skill);
    })
    .sort((a, b) => TOOLS[b.itemId].bonus - TOOLS[a.itemId].bonus)[0] ?? null;
}
